idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/WrapWithSafeLetCall.kt

```diff
@@ -0,0 +1,277 @@
+    }
+
+    val forUnsafeImplicitInvokeCall = diagnosticFixFactory<KtFirDiagnostic.UnsafeImplicitInvokeCall> { diagnostic ->
+        val callExpression = diagnostic.psi.parentOfType<KtCallExpression>(withSelf = true) ?: return@diagnosticFixFactory emptyList()
+        createWrapWithSafeLetCallInputForNullableExpression(
+            callExpression.calleeExpression,
+            isImplicitInvokeCallToMemberProperty = isCallingMemberProperty(callExpression)
+        )
+    }
+
+    private fun KtAnalysisSession.isCallingMemberProperty(callExpression: KtCallExpression): Boolean {
+        val calleeName = callExpression.calleeExpression?.text ?: return false
+        val propertySymbol = (callExpression.resolveCall() as? KtVariableWithInvokeFunctionCall)?.target ?: return false
+        val qualifiedExpression = callExpression.parent as? KtQualifiedExpression ?: return false
+        val scope = callExpression.containingKtFile.getScopeContextForPosition(qualifiedExpression)
+        // If no symbol in the local scope contains the called symbol, then the symbol must be a member symbol.
+        return scope.scopes.getCallableSymbols { it.identifierOrNullIfSpecial == calleeName }.all { it != propertySymbol }
+    }
+
+    val forUnsafeInfixCall = diagnosticFixFactory<KtFirDiagnostic.UnsafeInfixCall> { diagnostic ->
+        createWrapWithSafeLetCallInputForNullableExpressionIfMoreThanImmediateParentIsWrapped(diagnostic.receiverExpression)
+    }
+
+    val forUnsafeOperatorCall = diagnosticFixFactory<KtFirDiagnostic.UnsafeOperatorCall> { diagnostic ->
+        createWrapWithSafeLetCallInputForNullableExpressionIfMoreThanImmediateParentIsWrapped(diagnostic.receiverExpression)
+    }
+
+    val forArgumentTypeMismatch = diagnosticFixFactory<KtFirDiagnostic.ArgumentTypeMismatch> { diagnostic ->
+        if (diagnostic.isMismatchDueToNullability) createWrapWithSafeLetCallInputForNullableExpression(diagnostic.psi.wrappingExpressionOrSelf)
+        else emptyList()
+    }
+
+    private fun KtAnalysisSession.createWrapWithSafeLetCallInputForNullableExpressionIfMoreThanImmediateParentIsWrapped(
+        nullableExpression: KtExpression?,
+        isImplicitInvokeCallToMemberProperty: Boolean = false,
+    ): List<IntentionAction> {
+        val surroundingExpression = nullableExpression?.typableParentExpression
+        if (
+            surroundingExpression == null ||
+            // If the surrounding expression is at a place that accepts null value, then we don't provide wrap with let call because the
+            // plain safe call operator (?.) is a better fix.
+            isExpressionAtNullablePosition(surroundingExpression) ||
+            // In addition, if there is no parent that is at a nullable position, then we don't offer wrapping with let either because
+            // it still doesn't fix the code. Hence, the plain safe call operator is a better fix.
+            findParentExpressionAtNullablePosition(nullableExpression) == null
```

Comment:
I don't think it's equivalent if we remove `isExpressionAtNullablePosition`. Because without the check above, we would be offering the wrap-with-let quickfix when the safe call operator is better.

Reply:
Ah, I missed that subtle difference. Thanks!